linkerd repair segfault in 2.10.0

## Bug Report

### What is the issue?

```text
$ linkerd repair
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x17d910a]

goroutine 1 [running]:
github.com/linkerd/linkerd2/cli/cmd.resetVersion(0xc000240500, 0x1210, 0x1300)
	/linkerd-build/cli/cmd/repair.go:123 +0x7a
github.com/linkerd/linkerd2/cli/cmd.newCmdRepair.func1(0xc00045ab00, 0x2d55fc8, 0x0, 0x0, 0x0, 0x0)
	/linkerd-build/cli/cmd/repair.go:65 +0x469
github.com/spf13/cobra.(*Command).execute(0xc00045ab00, 0x2d55fc8, 0x0, 0x0, 0xc00045ab00, 0x2d55fc8)
	/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:842 +0x453
github.com/spf13/cobra.(*Command).ExecuteC(0x2d092a0, 0xc00018c010, 0x1, 0x1)
	/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:950 +0x349
github.com/spf13/cobra.(*Command).Execute(...)
	/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:887
main.main()
	/linkerd-build/cli/main.go:34 +0x19f
```

### How can it be reproduced?

Running the repair command while upgrading a cluster from 2.9.2 to 2.10.0

#### `linkerd check` output

```text
kubernetes-api
--------------
√ can initialize the client
√ can query the Kubernetes API

kubernetes-version
------------------
√ is running the minimum Kubernetes API version
√ is running the minimum kubectl version

linkerd-existence
-----------------
√ 'linkerd-config' config map exists
√ heartbeat ServiceAccount exist
√ control plane replica sets are ready
√ no unschedulable pods
√ controller pod is running

linkerd-config
--------------
√ control plane Namespace exists
√ control plane ClusterRoles exist
√ control plane ClusterRoleBindings exist
√ control plane ServiceAccounts exist
√ control plane CustomResourceDefinitions exist
√ control plane MutatingWebhookConfigurations exist
√ control plane ValidatingWebhookConfigurations exist
√ control plane PodSecurityPolicies exist

linkerd-identity
----------------
√ certificate config is valid
√ trust anchors are using supported crypto algorithm
√ trust anchors are within their validity period
√ trust anchors are valid for at least 60 days
√ issuer cert is using supported crypto algorithm
√ issuer cert is within its validity period
√ issuer cert is valid for at least 60 days
√ issuer cert is issued by the trust anchor

linkerd-webhooks-and-apisvc-tls
-------------------------------
√ proxy-injector webhook has valid cert
√ proxy-injector cert is valid for at least 60 days
√ sp-validator webhook has valid cert
√ sp-validator cert is valid for at least 60 days

linkerd-api
-----------
√ control plane pods are ready
√ can initialize the client
√ can query the control plane API

linkerd-version
---------------
√ can determine the latest version
√ cli is up-to-date

control-plane-version
---------------------
‼ control plane is up-to-date
    is running version 2.9.2 but the latest stable version is 2.10.0
    see https://linkerd.io/checks/#l5d-version-control for hints
‼ control plane and cli versions match
    control plane running stable-2.9.2 but cli running stable-2.10.0
    see https://linkerd.io/checks/#l5d-version-control for hints
```

### Environment

- Kubernetes Version: v1.17.11
- Cluster Environment: kops 
- Host OS: Debian stretch
- Linkerd version: 2.10.0

### Possible solution

Using linkerd version 2.9.4, the repair command works.

### Additional context

Non-helm.

```text
$ linkerd version
Client version: stable-2.10.0
Server version: stable-2.9.2
```

--- a/cli/cmd/repair.go
+++ b/cli/cmd/repair.go
@@ -25,9 +25,8 @@ var (
 	// 2.8 and below
 	repairNotApplicableVersionRegex = regexp.MustCompile(`^stable-2\.[0-8]\.([0-9]+)$`)
 
-	// repairApplicableVersionRegex matches versions 2.9.4 and above where repair
-	// would be useful
-	repairApplicableVersionRegex = regexp.MustCompile(`^stable-2\.(9|[0-9]{2})\.([0-9]+)$`)
+	// repairApplicableVersionRegex matches versions 2.9 versions upto 2.9.4
+	repairApplicableVersionRegex = regexp.MustCompile(`^stable-2\.9\.[0-4]$`)
 )
 
 // newCmdRepair creates a new cobra command `repair` which re-creates the
@@ -93,7 +92,7 @@ func repair(ctx context.Context, forced bool) error {
 			return fmt.Errorf("repair command is only applicable to 2.9 control-plane versions. Please try upgrading to the latest supported versions of Linkerd i.e 2.9.4 and above")
 		}
 
-		// Suggest 2.9.4 CLI version for all 2.9 server versions
+		// Suggest 2.9.4 CLI version for all 2.9 server versions upto 2.9.4
 		if repairApplicableVersionRegex.Match([]byte(serverVersion)) {
 			return fmt.Errorf("Please run the repair command with a `stable-2.9.4` CLI.\nRun `export LINKERD2_VERSION=\"stable-2.9.4\"; curl -sL https://run.linkerd.io/install | sh; unset LINKERD2_VERSION` to install the server version of the CLI")
 		}

--- a/cli/cmd/repair.go
+++ b/cli/cmd/repair.go
@@ -49,7 +49,7 @@ linkerd upgrade.`,
 
 			// Check if the CLI version matches with that of the server
 			clientVersion := version.Version
-			serverVersion := defaultVersionString
+			var serverVersion string
 			publicClient, err := public.NewExternalClient(cmd.Context(), controlPlaneNamespace, k8sAPI)
 			if err != nil {
 				return err

--- a/cli/cmd/repair.go
+++ b/cli/cmd/repair.go
@@ -21,9 +21,13 @@ import (
 )
 
 var (
-	repairNotApplicableVersionRegex = regexp.MustCompile("stable-2.[0-8].*")
+	// repairNotApplicableVersionRegex matches older versions of Linkerd i.e
+	// 2.8 and below
+	repairNotApplicableVersionRegex = regexp.MustCompile(`^stable-2\.[0-8]\.([0-9]+)$`)
 
-	repairApplicableVersionRegex = regexp.MustCompile("stable-2.[9].*")
+	// repairApplicableVersionRegex matches versions 2.9.4 and above where repair
+	// would be useful
+	repairApplicableVersionRegex = regexp.MustCompile(`^stable-2\.(9|[0-9]{2})\.([0-9]+)$`)
 )
 
 // newCmdRepair creates a new cobra command `repair` which re-creates the
@@ -93,11 +97,11 @@ func repair(ctx context.Context, forced bool) error {
 
 			// Suggest 2.9.4 CLI version for all 2.9 server versions
 			if repairApplicableVersionRegex.Match([]byte(serverVersion)) {
-				return fmt.Errorf("Please run the repair command with a `2.9.4` CLI.\nRun `LINKERD2_VERSION=\"stable-2.9.4\"; curl -sL https://run.linkerd.io/install | sh` to install the server version of the CLI")
+				return fmt.Errorf("Please run the repair command with a `2.9.4` CLI.\nRun `curl -sL https://run.linkerd.io/install | LINKERD2_VERSION=\"2.9.4\" sh` to install the server version of the CLI")
 			}
 
 			// Suggest server version for everything else. This includes all edge versions
-			return fmt.Errorf("Please run the repair command with a CLI that has the same version as the control plane.\nRun `LINKERD2_VERSION=\"%s\"; curl -sL https://run.linkerd.io/install | sh` to install the server version of the CLI", serverVersion)
+			return fmt.Errorf("Please run the repair command with a CLI that has the same version as the control plane.\nRun `curl -sL https://run.linkerd.io/install | LINKERD2_VERSION=\"%s\" sh` to install the server version of the CLI", serverVersion)
 		}
 	}
 

--- a/cli/cmd/repair.go
+++ b/cli/cmd/repair.go
@@ -3,12 +3,16 @@ package cmd
 import (
 	"errors"
 	"fmt"
+	"regexp"
 	"time"
 
 	"github.com/golang/protobuf/ptypes"
+	"github.com/linkerd/linkerd2/controller/api/public"
 	pb "github.com/linkerd/linkerd2/controller/gen/config"
 	"github.com/linkerd/linkerd2/pkg/charts/linkerd2"
+	"github.com/linkerd/linkerd2/pkg/healthcheck"
 	"github.com/linkerd/linkerd2/pkg/k8s"
+	"github.com/linkerd/linkerd2/pkg/version"
 	"github.com/spf13/cobra"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"sigs.k8s.io/yaml"
@@ -41,6 +45,31 @@ linkerd upgrade.`,
 			_, err = k8sAPI.CoreV1().Secrets(controlPlaneNamespace).Get(cmd.Context(), "linkerd-config-overrides", metav1.GetOptions{})
 			if err == nil {
 				return errors.New("secret/linkerd-config-overrides already exists. If you need to regenerate this resource, please delete it before proceeding")
+			}
+
+			// Check if the CLI version matches with that of the server
+			clientVersion := version.Version
+			serverVersion := defaultVersionString
+			publicClient, err := public.NewExternalClient(cmd.Context(), controlPlaneNamespace, k8sAPI)
+			if err != nil {
+				return err
+			}
+			serverVersion, err = healthcheck.GetServerVersion(cmd.Context(), publicClient)
+			if err != nil {
+				return err
+			}
+
+			if serverVersion != clientVersion {
+				twoPointEightOrNineRegex, err := regexp.Compile("stable-2.[8-9].*")
+				if err != nil {
+					return err
+				}
+				helperVersion := serverVersion
+				// Use 2.9.4 CLI version for all 2.9 versions
+				if twoPointEightOrNineRegex.Match([]byte(serverVersion)) {
+					helperVersion = "stable-2.9.4"
+				}
+				return fmt.Errorf("Please run the repair command with a CLI version as that of the server i.e %s\nRun `LINKERD2_VERSION=\"%s\"; curl -sL https://run.linkerd.io/install | sh` to install the server version of the CLI", serverVersion, helperVersion)
 			}
 
 			// Load the stored config

--- a/cli/cmd/repair.go
+++ b/cli/cmd/repair.go
@@ -97,7 +97,7 @@ func repair(ctx context.Context, forced bool) error {
 
 			// Suggest 2.9.4 CLI version for all 2.9 server versions
 			if repairApplicableVersionRegex.Match([]byte(serverVersion)) {
-				return fmt.Errorf("Please run the repair command with a `2.9.4` CLI.\nRun `export LINKERD2_VERSION=\"stable-2.9.4\"; curl -sL https://run.linkerd.io/install | sh; unset LINKERD2_VERSION` to install the server version of the CLI")
+				return fmt.Errorf("Please run the repair command with a `stable-2.9.4` CLI.\nRun `export LINKERD2_VERSION=\"stable-2.9.4\"; curl -sL https://run.linkerd.io/install | sh; unset LINKERD2_VERSION` to install the server version of the CLI")
 			}
 
 			// Suggest server version for everything else. This includes all edge versions
@@ -177,8 +177,25 @@ func resetVersion(values *linkerd2.Values) error {
 	if err != nil {
 		return err
 	}
-	values.DebugContainer.Image.Version = defaults.DebugContainer.Image.Version
-	values.Proxy.Image.Version = defaults.Proxy.Image.Version
+
+	if values.DebugContainer != nil {
+		if values.DebugContainer.Image != nil {
+			values.DebugContainer.Image.Version = defaults.DebugContainer.Image.Version
+		}
+	}
+
+	if values.Proxy != nil {
+		if values.Proxy.Image != nil {
+			values.Proxy.Image.Version = defaults.Proxy.Image.Version
+		}
+	}
+
+	if values.ProxyInit != nil {
+		if values.ProxyInit.Image != nil {
+			values.ProxyInit.Image.Version = defaults.ProxyInit.Image.Version
+		}
+	}
+
 	values.CliVersion = defaults.CliVersion
 	values.ControllerImageVersion = defaults.ControllerImageVersion
 	values.LinkerdVersion = defaults.LinkerdVersion

--- a/cli/cmd/repair.go
+++ b/cli/cmd/repair.go
@@ -28,6 +28,7 @@ var (
 // linkerd-config-overrides secret if it has been deleted.
 func newCmdRepair() *cobra.Command {
 
+	var force bool
 	cmd := &cobra.Command{
 		Use:   "repair",
 		Short: "Output the secret/linkerd-config-overrides resource if it has been deleted",
@@ -42,7 +43,7 @@ linkerd upgrade.`,
 		Example: "  linkerd repair | kubectl apply -f -",
 		Args:    cobra.NoArgs,
 		RunE: func(cmd *cobra.Command, args []string) error {
-			err := repair(cmd.Context())
+			err := repair(cmd.Context(), force)
 			if err != nil {
 				fmt.Fprintf(os.Stderr, err.Error()+"\n")
 				os.Exit(1)
@@ -51,10 +52,12 @@ linkerd upgrade.`,
 		},
 	}
 
+	cmd.Flags().BoolVarP(&force, "force", "f", force, "Force render even if the CLI and control-plane versions don't match")
+
 	return cmd
 }
 
-func repair(ctx context.Context) error {
+func repair(ctx context.Context, forced bool) error {
 	k8sAPI, err := k8s.NewAPI(kubeconfigPath, kubeContext, impersonate, impersonateGroup, 0)
 	if err != nil {
 		return err
@@ -78,13 +81,15 @@ func repair(ctx context.Context) error {
 		return err
 	}
 
-	if serverVersion != clientVersion {
-		helperVersion := serverVersion
-		// Use 2.9.4 CLI version for all 2.9 versions
-		if versionRegex.Match([]byte(serverVersion)) {
-			helperVersion = "stable-2.9.4"
+	if !forced {
+		if serverVersion != clientVersion {
+			helperVersion := serverVersion
+			// Use 2.9.4 CLI version for all 2.9 versions
+			if versionRegex.Match([]byte(serverVersion)) {
+				helperVersion = "stable-2.9.4"
+			}
+			return fmt.Errorf("Please run the repair command with a CLI that has the same version as the control plane.\nRun `LINKERD2_VERSION=\"%s\"; curl -sL https://run.linkerd.io/install | sh` to install the server version of the CLI", helperVersion)
 		}
-		return fmt.Errorf("Please run the repair command with a CLI that has the same version as the control plane.\nRun `LINKERD2_VERSION=\"%s\"; curl -sL https://run.linkerd.io/install | sh` to install the server version of the CLI", helperVersion)
 	}
 
 	// Load the stored config

--- a/cli/cmd/repair.go
+++ b/cli/cmd/repair.go
@@ -87,22 +87,19 @@ func repair(ctx context.Context, forced bool) error {
 		return err
 	}
 
-	if !forced {
-		if serverVersion != clientVersion {
-
-			// Suggest directly upgrading to 2.9.4 or above for older versions
-			if repairNotApplicableVersionRegex.Match([]byte(serverVersion)) {
-				return fmt.Errorf("repair command is only applicable to 2.9 control-plane versions. Please try upgrading to the latest supported versions of Linkerd i.e 2.9.4 and above")
-			}
-
-			// Suggest 2.9.4 CLI version for all 2.9 server versions
-			if repairApplicableVersionRegex.Match([]byte(serverVersion)) {
-				return fmt.Errorf("Please run the repair command with a `stable-2.9.4` CLI.\nRun `export LINKERD2_VERSION=\"stable-2.9.4\"; curl -sL https://run.linkerd.io/install | sh; unset LINKERD2_VERSION` to install the server version of the CLI")
-			}
-
-			// Suggest server version for everything else. This includes all edge versions
-			return fmt.Errorf("Please run the repair command with a CLI that has the same version as the control plane.\nRun `export LINKERD2_VERSION=\"%s\"; curl -sL https://run.linkerd.io/install | sh; unset LINKERD2_VERSION` to install the server version of the CLI", serverVersion)
+	if !forced && serverVersion != clientVersion {
+		// Suggest directly upgrading to 2.9.4 or above for older versions
+		if repairNotApplicableVersionRegex.Match([]byte(serverVersion)) {
+			return fmt.Errorf("repair command is only applicable to 2.9 control-plane versions. Please try upgrading to the latest supported versions of Linkerd i.e 2.9.4 and above")
 		}
+
+		// Suggest 2.9.4 CLI version for all 2.9 server versions
+		if repairApplicableVersionRegex.Match([]byte(serverVersion)) {
+			return fmt.Errorf("Please run the repair command with a `stable-2.9.4` CLI.\nRun `export LINKERD2_VERSION=\"stable-2.9.4\"; curl -sL https://run.linkerd.io/install | sh; unset LINKERD2_VERSION` to install the server version of the CLI")
+		}
+
+		// Suggest server version for everything else. This includes all edge versions
+		return fmt.Errorf("Please run the repair command with a CLI that has the same version as the control plane.\nRun `export LINKERD2_VERSION=\"%s\"; curl -sL https://run.linkerd.io/install | sh; unset LINKERD2_VERSION` to install the server version of the CLI", serverVersion)
 	}
 
 	// Load the stored config
@@ -123,7 +120,6 @@ func repair(ctx context.Context, forced bool) error {
 		return fmt.Errorf("Failed to load values from linkerd-config: %s", err)
 	}
 
-	// Reset version fields
 	err = resetVersion(&values)
 	if err != nil {
 		return fmt.Errorf("Failed to reset version fields in linkerd-config: %s", err)
@@ -178,22 +174,16 @@ func resetVersion(values *linkerd2.Values) error {
 		return err
 	}
 
-	if values.DebugContainer != nil {
-		if values.DebugContainer.Image != nil {
-			values.DebugContainer.Image.Version = defaults.DebugContainer.Image.Version
-		}
+	if values.DebugContainer != nil && values.DebugContainer.Image != nil {
+		values.DebugContainer.Image.Version = defaults.DebugContainer.Image.Version
 	}
 
-	if values.Proxy != nil {
-		if values.Proxy.Image != nil {
-			values.Proxy.Image.Version = defaults.Proxy.Image.Version
-		}
+	if values.Proxy != nil && values.Proxy.Image != nil {
+		values.Proxy.Image.Version = defaults.Proxy.Image.Version
 	}
 
-	if values.ProxyInit != nil {
-		if values.ProxyInit.Image != nil {
-			values.ProxyInit.Image.Version = defaults.ProxyInit.Image.Version
-		}
+	if values.ProxyInit != nil && values.ProxyInit.Image != nil {
+		values.ProxyInit.Image.Version = defaults.ProxyInit.Image.Version
 	}
 
 	values.CliVersion = defaults.CliVersion

--- a/cli/cmd/repair.go
+++ b/cli/cmd/repair.go
@@ -97,11 +97,11 @@ func repair(ctx context.Context, forced bool) error {
 
 			// Suggest 2.9.4 CLI version for all 2.9 server versions
 			if repairApplicableVersionRegex.Match([]byte(serverVersion)) {
-				return fmt.Errorf("Please run the repair command with a `2.9.4` CLI.\nRun `curl -sL https://run.linkerd.io/install | LINKERD2_VERSION=\"2.9.4\" sh` to install the server version of the CLI")
+				return fmt.Errorf("Please run the repair command with a `2.9.4` CLI.\nRun `export LINKERD2_VERSION=\"stable-2.9.4\"; curl -sL https://run.linkerd.io/install | sh; unset LINKERD2_VERSION` to install the server version of the CLI")
 			}
 
 			// Suggest server version for everything else. This includes all edge versions
-			return fmt.Errorf("Please run the repair command with a CLI that has the same version as the control plane.\nRun `curl -sL https://run.linkerd.io/install | LINKERD2_VERSION=\"%s\" sh` to install the server version of the CLI", serverVersion)
+			return fmt.Errorf("Please run the repair command with a CLI that has the same version as the control plane.\nRun `export LINKERD2_VERSION=\"%s\"; curl -sL https://run.linkerd.io/install | sh; unset LINKERD2_VERSION` to install the server version of the CLI", serverVersion)
 		}
 	}
 

--- a/cli/cmd/repair.go
+++ b/cli/cmd/repair.go
@@ -1,8 +1,10 @@
 package cmd
 
 import (
+	"context"
 	"errors"
 	"fmt"
+	"os"
 	"regexp"
 	"time"
 
@@ -40,101 +42,110 @@ linkerd upgrade.`,
 		Example: "  linkerd repair | kubectl apply -f -",
 		Args:    cobra.NoArgs,
 		RunE: func(cmd *cobra.Command, args []string) error {
-			k8sAPI, err := k8s.NewAPI(kubeconfigPath, kubeContext, impersonate, impersonateGroup, 0)
+			err := repair(cmd.Context())
 			if err != nil {
-				return err
+				fmt.Fprintf(os.Stderr, err.Error()+"\n")
+				os.Exit(1)
 			}
-
-			// Error out if linkerd-config-overrides already exists.
-			_, err = k8sAPI.CoreV1().Secrets(controlPlaneNamespace).Get(cmd.Context(), "linkerd-config-overrides", metav1.GetOptions{})
-			if err == nil {
-				return errors.New("secret/linkerd-config-overrides already exists. If you need to regenerate this resource, please delete it before proceeding")
-			}
-
-			// Check if the CLI version matches with that of the server
-			clientVersion := version.Version
-			var serverVersion string
-			publicClient, err := public.NewExternalClient(cmd.Context(), controlPlaneNamespace, k8sAPI)
-			if err != nil {
-				return err
-			}
-			serverVersion, err = healthcheck.GetServerVersion(cmd.Context(), publicClient)
-			if err != nil {
-				return err
-			}
-
-			if serverVersion != clientVersion {
-				helperVersion := serverVersion
-				// Use 2.9.4 CLI version for all 2.9 versions
-				if versionRegex.Match([]byte(serverVersion)) {
-					helperVersion = "stable-2.9.4"
-				}
-				return fmt.Errorf("Please run the repair command with a CLI that has the same version as the control plane.\nRun `LINKERD2_VERSION=\"%s\"; curl -sL https://run.linkerd.io/install | sh` to install the server version of the CLI", helperVersion)
-			}
-
-			// Load the stored config
-			config, err := k8sAPI.CoreV1().ConfigMaps(controlPlaneNamespace).Get(cmd.Context(), "linkerd-config", metav1.GetOptions{})
-			if err != nil {
-				return fmt.Errorf("Failed to load linkerd-config: %s", err)
-			}
-
-			values := linkerd2.Values{}
-
-			valuesRaw, ok := config.Data["values"]
-			if !ok {
-				return errors.New("values not found in linkerd-config")
-			}
-
-			err = yaml.Unmarshal([]byte(valuesRaw), &values)
-			if err != nil {
-				return fmt.Errorf("Failed to load values from linkerd-config: %s", err)
-			}
-
-			// Reset version fields
-			err = resetVersion(&values)
-			if err != nil {
-				return fmt.Errorf("Failed to reset version fields in linkerd-config: %s", err)
-			}
-
-			clockSkewDuration, err := time.ParseDuration(values.Identity.Issuer.ClockSkewAllowance)
-			if err != nil {
-				return fmt.Errorf("Failed to parse ClockSkewAllowance from linkerd-config: %s", err)
-			}
-			issuanceLifetime, err := time.ParseDuration(values.Identity.Issuer.IssuanceLifetime)
-			if err != nil {
-				return fmt.Errorf("Failed to parse IssuanceLifetime from linkerd-config: %s", err)
-			}
-			idCtx := pb.IdentityContext{
-				TrustAnchorsPem:    values.IdentityTrustAnchorsPEM,
-				Scheme:             values.Identity.Issuer.Scheme,
-				ClockSkewAllowance: ptypes.DurationProto(clockSkewDuration),
-				IssuanceLifetime:   ptypes.DurationProto(issuanceLifetime),
-				TrustDomain:        values.IdentityTrustDomain,
-			}
-
-			// Populate identity values
-			err = fetchIdentityValues(cmd.Context(), k8sAPI, &idCtx, &values)
-			if err != nil {
-				return fmt.Errorf("Failed to load issuer credentials: %s", err)
-			}
-
-			valuesMap, err := values.ToMap()
-			if err != nil {
-				return fmt.Errorf("Failed to convert Values into a map: %s", err)
-			}
-
-			overrides, err := renderOverrides(valuesMap, true)
-			if err != nil {
-				return fmt.Errorf("Failed to render overrides: %s", err)
-			}
-
-			fmt.Print(string(overrides))
-
 			return nil
 		},
 	}
 
 	return cmd
+}
+
+func repair(ctx context.Context) error {
+	k8sAPI, err := k8s.NewAPI(kubeconfigPath, kubeContext, impersonate, impersonateGroup, 0)
+	if err != nil {
+		return err
+	}
+
+	// Error out if linkerd-config-overrides already exists.
+	_, err = k8sAPI.CoreV1().Secrets(controlPlaneNamespace).Get(ctx, "linkerd-config-overrides", metav1.GetOptions{})
+	if err == nil {
+		return errors.New("secret/linkerd-config-overrides already exists. If you need to regenerate this resource, please delete it before proceeding")
+	}
+
+	// Check if the CLI version matches with that of the server
+	clientVersion := version.Version
+	var serverVersion string
+	publicClient, err := public.NewExternalClient(ctx, controlPlaneNamespace, k8sAPI)
+	if err != nil {
+		return err
+	}
+	serverVersion, err = healthcheck.GetServerVersion(ctx, publicClient)
+	if err != nil {
+		return err
+	}
+
+	if serverVersion != clientVersion {
+		helperVersion := serverVersion
+		// Use 2.9.4 CLI version for all 2.9 versions
+		if versionRegex.Match([]byte(serverVersion)) {
+			helperVersion = "stable-2.9.4"
+		}
+		return fmt.Errorf("Please run the repair command with a CLI that has the same version as the control plane.\nRun `LINKERD2_VERSION=\"%s\"; curl -sL https://run.linkerd.io/install | sh` to install the server version of the CLI", helperVersion)
+	}
+
+	// Load the stored config
+	config, err := k8sAPI.CoreV1().ConfigMaps(controlPlaneNamespace).Get(ctx, "linkerd-config", metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("Failed to load linkerd-config: %s", err)
+	}
+
+	values := linkerd2.Values{}
+
+	valuesRaw, ok := config.Data["values"]
+	if !ok {
+		return errors.New("values not found in linkerd-config")
+	}
+
+	err = yaml.Unmarshal([]byte(valuesRaw), &values)
+	if err != nil {
+		return fmt.Errorf("Failed to load values from linkerd-config: %s", err)
+	}
+
+	// Reset version fields
+	err = resetVersion(&values)
+	if err != nil {
+		return fmt.Errorf("Failed to reset version fields in linkerd-config: %s", err)
+	}
+
+	clockSkewDuration, err := time.ParseDuration(values.Identity.Issuer.ClockSkewAllowance)
+	if err != nil {
+		return fmt.Errorf("Failed to parse ClockSkewAllowance from linkerd-config: %s", err)
+	}
+	issuanceLifetime, err := time.ParseDuration(values.Identity.Issuer.IssuanceLifetime)
+	if err != nil {
+		return fmt.Errorf("Failed to parse IssuanceLifetime from linkerd-config: %s", err)
+	}
+	idCtx := pb.IdentityContext{
+		TrustAnchorsPem:    values.IdentityTrustAnchorsPEM,
+		Scheme:             values.Identity.Issuer.Scheme,
+		ClockSkewAllowance: ptypes.DurationProto(clockSkewDuration),
+		IssuanceLifetime:   ptypes.DurationProto(issuanceLifetime),
+		TrustDomain:        values.IdentityTrustDomain,
+	}
+
+	// Populate identity values
+	err = fetchIdentityValues(ctx, k8sAPI, &idCtx, &values)
+	if err != nil {
+		return fmt.Errorf("Failed to load issuer credentials: %s", err)
+	}
+
+	valuesMap, err := values.ToMap()
+	if err != nil {
+		return fmt.Errorf("Failed to convert Values into a map: %s", err)
+	}
+
+	overrides, err := renderOverrides(valuesMap, true)
+	if err != nil {
+		return fmt.Errorf("Failed to render overrides: %s", err)
+	}
+
+	fmt.Print(string(overrides))
+
+	return nil
 }
 
 // resetVersion sets all linkerd version fields to the chart's default version

--- a/cli/cmd/repair.go
+++ b/cli/cmd/repair.go
@@ -18,6 +18,10 @@ import (
 	"sigs.k8s.io/yaml"
 )
 
+var (
+	versionRegex = regexp.MustCompile("stable-2.[8-9].*")
+)
+
 // newCmdRepair creates a new cobra command `repair` which re-creates the
 // linkerd-config-overrides secret if it has been deleted.
 func newCmdRepair() *cobra.Command {
@@ -60,10 +64,6 @@ linkerd upgrade.`,
 			}
 
 			if serverVersion != clientVersion {
-				versionRegex, err := regexp.Compile("stable-2.[8-9].*")
-				if err != nil {
-					return err
-				}
 				helperVersion := serverVersion
 				// Use 2.9.4 CLI version for all 2.9 versions
 				if versionRegex.Match([]byte(serverVersion)) {

--- a/cli/cmd/repair.go
+++ b/cli/cmd/repair.go
@@ -94,11 +94,11 @@ func repair(ctx context.Context, forced bool) error {
 
 		// Suggest 2.9.4 CLI version for all 2.9 server versions upto 2.9.4
 		if repairApplicableVersionRegex.Match([]byte(serverVersion)) {
-			return fmt.Errorf("Please run the repair command with a `stable-2.9.4` CLI.\nRun `export LINKERD2_VERSION=\"stable-2.9.4\"; curl -sL https://run.linkerd.io/install | sh; unset LINKERD2_VERSION` to install the server version of the CLI")
+			return fmt.Errorf("Please run the repair command with a `stable-2.9.4` CLI.\nRun `curl -sL https://run.linkerd.io/install | LINKERD2_VERSION=\"stable-2.9.4\" sh` to install the server version of the CLI")
 		}
 
 		// Suggest server version for everything else. This includes all edge versions
-		return fmt.Errorf("Please run the repair command with a CLI that has the same version as the control plane.\nRun `export LINKERD2_VERSION=\"%s\"; curl -sL https://run.linkerd.io/install | sh; unset LINKERD2_VERSION` to install the server version of the CLI", serverVersion)
+		return fmt.Errorf("Please run the repair command with a CLI that has the same version as the control plane.\nRun `curl -sL https://run.linkerd.io/install | LINKERD2_VERSION=\"%s\" sh` to install the server version of the CLI", serverVersion)
 	}
 
 	// Load the stored config

--- a/cli/cmd/repair.go
+++ b/cli/cmd/repair.go
@@ -60,16 +60,16 @@ linkerd upgrade.`,
 			}
 
 			if serverVersion != clientVersion {
-				twoPointEightOrNineRegex, err := regexp.Compile("stable-2.[8-9].*")
+				versionRegex, err := regexp.Compile("stable-2.[8-9].*")
 				if err != nil {
 					return err
 				}
 				helperVersion := serverVersion
 				// Use 2.9.4 CLI version for all 2.9 versions
-				if twoPointEightOrNineRegex.Match([]byte(serverVersion)) {
+				if versionRegex.Match([]byte(serverVersion)) {
 					helperVersion = "stable-2.9.4"
 				}
-				return fmt.Errorf("Please run the repair command with a CLI version as that of the server.\nRun `LINKERD2_VERSION=\"%s\"; curl -sL https://run.linkerd.io/install | sh` to install the server version of the CLI", helperVersion)
+				return fmt.Errorf("Please run the repair command with a CLI that has the same version as the control plane.\nRun `LINKERD2_VERSION=\"%s\"; curl -sL https://run.linkerd.io/install | sh` to install the server version of the CLI", helperVersion)
 			}
 
 			// Load the stored config

--- a/cli/cmd/repair.go
+++ b/cli/cmd/repair.go
@@ -21,7 +21,9 @@ import (
 )
 
 var (
-	versionRegex = regexp.MustCompile("stable-2.[8-9].*")
+	repairNotApplicableVersionRegex = regexp.MustCompile("stable-2.[0-8].*")
+
+	repairApplicableVersionRegex = regexp.MustCompile("stable-2.[9].*")
 )
 
 // newCmdRepair creates a new cobra command `repair` which re-creates the
@@ -84,8 +86,14 @@ func repair(ctx context.Context, forced bool) error {
 	if !forced {
 		if serverVersion != clientVersion {
 			helperVersion := serverVersion
+
+			// Suggest directly upgrading to 2.9.4 or above for older versions
+			if repairNotApplicableVersionRegex.Match([]byte(serverVersion)) {
+				return fmt.Errorf("repair command is only applicable to 2.9 control-plane versions. Please try upgrading to the latest supported versions of Linkerd i.e 2.9.4 and above")
+			}
+
 			// Use 2.9.4 CLI version for all 2.9 versions
-			if versionRegex.Match([]byte(serverVersion)) {
+			if repairApplicableVersionRegex.Match([]byte(serverVersion)) {
 				helperVersion = "stable-2.9.4"
 			}
 			return fmt.Errorf("Please run the repair command with a CLI that has the same version as the control plane.\nRun `LINKERD2_VERSION=\"%s\"; curl -sL https://run.linkerd.io/install | sh` to install the server version of the CLI", helperVersion)

--- a/cli/cmd/repair.go
+++ b/cli/cmd/repair.go
@@ -69,7 +69,7 @@ linkerd upgrade.`,
 				if twoPointEightOrNineRegex.Match([]byte(serverVersion)) {
 					helperVersion = "stable-2.9.4"
 				}
-				return fmt.Errorf("Please run the repair command with a CLI version as that of the server i.e %s\nRun `LINKERD2_VERSION=\"%s\"; curl -sL https://run.linkerd.io/install | sh` to install the server version of the CLI", serverVersion, helperVersion)
+				return fmt.Errorf("Please run the repair command with a CLI version as that of the server.\nRun `LINKERD2_VERSION=\"%s\"; curl -sL https://run.linkerd.io/install | sh` to install the server version of the CLI", helperVersion)
 			}
 
 			// Load the stored config

--- a/cli/cmd/repair.go
+++ b/cli/cmd/repair.go
@@ -85,18 +85,19 @@ func repair(ctx context.Context, forced bool) error {
 
 	if !forced {
 		if serverVersion != clientVersion {
-			helperVersion := serverVersion
 
 			// Suggest directly upgrading to 2.9.4 or above for older versions
 			if repairNotApplicableVersionRegex.Match([]byte(serverVersion)) {
 				return fmt.Errorf("repair command is only applicable to 2.9 control-plane versions. Please try upgrading to the latest supported versions of Linkerd i.e 2.9.4 and above")
 			}
 
-			// Use 2.9.4 CLI version for all 2.9 versions
+			// Suggest 2.9.4 CLI version for all 2.9 server versions
 			if repairApplicableVersionRegex.Match([]byte(serverVersion)) {
-				helperVersion = "stable-2.9.4"
+				return fmt.Errorf("Please run the repair command with a `2.9.4` CLI.\nRun `LINKERD2_VERSION=\"stable-2.9.4\"; curl -sL https://run.linkerd.io/install | sh` to install the server version of the CLI")
 			}
-			return fmt.Errorf("Please run the repair command with a CLI that has the same version as the control plane.\nRun `LINKERD2_VERSION=\"%s\"; curl -sL https://run.linkerd.io/install | sh` to install the server version of the CLI", helperVersion)
+
+			// Suggest server version for everything else. This includes all edge versions
+			return fmt.Errorf("Please run the repair command with a CLI that has the same version as the control plane.\nRun `LINKERD2_VERSION=\"%s\"; curl -sL https://run.linkerd.io/install | sh` to install the server version of the CLI", serverVersion)
 		}
 	}
 

